include/benchmark/benchmark_api.h

```diff
@@ -227,6 +234,113 @@ inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp const& value) {
 // FIXME Add ClobberMemory() for non-gnu compilers
 #endif
 
+
+
+class Counter {
+public:
+
+  typedef enum {
+    kDefaults       = 0x00,
+    /** Mark the counter as a rate. It will be presented divided by the duration of the benchmark. */
+    kIsRate         = 0x01 << 0,
+    /** Mark the counter as a thread-average quantity. It will be presented divided by the number of threads. */
+    kAvgThreads     = 0x01 << 1
+  } Flags_e;
+
+public:
+
+  Counter(std::string const& n, double v = 0., uint32_t flags = kDefaults);
+
+  std::string const& Name() const { return name_; }
+  double Value() const { return value_; }
+
+  void Set(double v) { value_ = v; }
+
+  Counter& operator+= (double v) { value_ += v; return *this; }
+  Counter& operator-= (double v) { value_ -= v; return *this; }
+  Counter& operator*= (double v) { value_ *= v; return *this; }
+  Counter& operator/= (double v) { value_ /= v; return *this; }
+
+  void Finish(double cpu_time, double num_threads);
+
+private:
+
+  std::string name_;
+  double      value_;
+  uint32_t    flags_;
+  // <jppm> maybe we could add a description field...
+
+};
+
+
+
+class BenchmarkCounters {
+public:
+
+  typedef Counter                         value_type;
+  typedef std::vector< Counter >          container_type;
+  typedef container_type::const_iterator  const_iterator;
+
+public:
+
+  BenchmarkCounters(size_t initial_capacity = 16);
+
+  // Add/Set a counter. Will overwrite any additional counter properties.
+  size_t  Set(Counter const& c);
+  // Add/Set a counter. When the counter already exists, f is ignored.
+  size_t  Set(const char*        n, double v, uint32_t f = Counter::kDefaults);
+  // Add/Set a counter. When the counter already exists, f is ignored.
+  size_t  Set(std::string const& n, double v, uint32_t f = Counter::kDefaults);
```

Comment:
I used std::string as per your suggestion in #240 . I acknowledge that does not mean that it should be exposed in the public header. Would you be OK with a PIMPL idiom here? That would hide the need for including both <vector> and <string> .